### PR TITLE
Add Contour Maps

### PR DESCRIPTION
**Pull request by AbelVM**
For new analysis use the following checklist:

- [x] Uses `the_geom geometry(Geometry, 4326)` column for all geospatial operations.
- [x] Outputs a `the_geom geometry(Geometry, 4326)` column.
- [x] Outputs a `cartodb_id numeric` column.
- [x] Uses `{cache: true}` option when it needs full knowledge of the table it. Hints: aggregations, window functions.
- [x] Uses `{cache: true}` if it access external services.
- [x] Naming uses a-z lowercase and hyphens.
- [x] All mandatory params cannot be made optional.
- [x] Avoids using CTEs for join operations when result is not cached.

### CODE REVIEW

**Comment by AbelVM** [comment]
cc @dgaubert , so we can review this PR next wednesday

File: lib/node/nodes/contour-map.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    // the scattered points dataset with real values
+    source: Node.PARAM.NODE(Node.GEOMETRY.POINT),
+    // column of values in source dataset
+    column: Node.PARAM.STRING(),
+    // buffer around my dataset
+    buffer: Node.PARAM.NULLABLE(Node.PARAM.NUMBER(), 0.2),
+    // 0: nearest neighbor, 1: barymetric, 2: IDW
+    method: Node.PARAM.NULLABLE(Node.PARAM.ENUM('nearest_neighbor', 'barymetric', 'IDW'), 'barymetric'),
+    // classification method
+    class_method: Node.PARAM.NULLABLE(Node.PARAM.ENUM('equals', 'heads_and_tails', 'jenks', 'quantiles'), 'quantiles'),
```
Comment (rochoa) [inline lib/node/nodes/contour-map.js]: I would rename head_and_tails to headstails to be coherent with cartodb-postgresql extension methods.

File: test/acceptance/contour.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        testHelper.createAnalyses(analysisDefinition, function (err, analysisResult) {
+            if (err) {
+                return done(err);
+            }
+
+            testHelper.getRows(analysisResult.getQuery(), function (err, rows) {
+                if (err) {
+                    return done(err);
+                }
+                assert.equal(rows.length, analysisDefinition.params.steps);
```
Comment (dgaubert) [inline test/acceptance/contour.js]: It'd be nice assert if `cartodb_id` and `the_geom` are in the analysis' output. Like we did in `gravity`

Comment (AbelVM) [inline test/acceptance/contour.js]: Okidoki

File: test/fixtures/cdb_crankshaft/cdb_kmeans.sql [hunk 3]
```diff
@@ -10,3 +12,38 @@ CREATE OR REPLACE FUNCTION cdb_crankshaft.CDB_KMeans(query text, no_clusters int
     RETURN;
     END;
 $$ LANGUAGE plpgsql;
+
+CREATE OR REPLACE FUNCTION cdb_crankshaft.CDB_Contour(
```
Comment (dgaubert) [inline test/fixtures/cdb_crankshaft/cdb_kmeans.sql]: Why not to use the same approach that we did in `gravity`?. I mean, create a separated file with `CDB_Cotour`

Comment (AbelVM) [inline test/fixtures/cdb_crankshaft/cdb_kmeans.sql]: okidoki 2

File: lib/node/nodes/contour.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+var PARAMS = {
+    // the scattered points dataset with real values
+    source: Node.PARAM.NODE(Node.GEOMETRY.POINT),
+    // column of values in source dataset
+    column: Node.PARAM.STRING(),
+    // buffer around my dataset
+    buffer: Node.PARAM.NULLABLE(Node.PARAM.NUMBER(), 0.2),
+    // 0: nearest neighbor, 1: barymetric, 2: IDW
+    method: Node.PARAM.NULLABLE(Node.PARAM.ENUM('nearest_neighbor', 'barymetric', 'IDW'), 'barymetric'),
```
Comment (dgaubert) [inline lib/node/nodes/contour.js]: We use these literals twice. We should create a `variable/constant` and use it instead.

File: lib/node/nodes/contour.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    // classification method
+    class_method: Node.PARAM.NULLABLE(Node.PARAM.ENUM('equals', 'headstails', 'jenks', 'quantiles'), 'quantiles'),
```
Comment (dgaubert) [inline lib/node/nodes/contour.js]: Same here. Create `variable/constant` for string literals

File: lib/node/nodes/contour.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    '        {{=it.steps}},',
+    '        {{=it.resolution}}) contour'
+].join('\n'));
+
+Contour.prototype.sql = function () {
+    return contourTemplate({
+        source_query: this.source.getQuery(),
+        column: this.column,
+        buffer: this.buffer,
+        method: ['nearest_neighbor', 'barymetric','IDW'].indexOf(this.method),
```
Comment (dgaubert) [inline lib/node/nodes/contour.js]: Same here

File: lib/node/nodes/contour.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        class: ['equals', 'headstails', 'jenks', 'quantiles'].indexOf(this.class_method),
```
Comment (dgaubert) [inline lib/node/nodes/contour.js]: And here

File: lib/node/nodes/contour.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    '    array_agg({{=it.column}}::numeric) as v',
+    '    FROM a',
+    '    WHERE {{=it.column}} is not null   ',
+    ')',
+    'SELECT',
+    '    contour.bin as cartodb_id,',
+    '    contour.*',
+    'FROM',
+    '    b,',
+    '    cdb_crankshaft.CDB_contour(b.g, b.v,',
```
Comment (dgaubert) [inline lib/node/nodes/contour.js]: Could we use meaningful names here? I don't know what `b.g`, `b.v` are...

File: lib/node/nodes/contour.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    // the scattered points dataset with real values
+    source: Node.PARAM.NODE(Node.GEOMETRY.POINT),
+    // column of values in source dataset
+    column: Node.PARAM.STRING(),
+    // buffer around my dataset
+    buffer: Node.PARAM.NULLABLE(Node.PARAM.NUMBER(), 0.2),
+    // 0: nearest neighbor, 1: barymetric, 2: IDW
+    method: Node.PARAM.NULLABLE(Node.PARAM.ENUM.apply(null,METHODS), METHODS[1]),
+    // classification method
+    class_method: Node.PARAM.NULLABLE(Node.PARAM.ENUM.apply(null,CLASS_METHODS), CLASS_METHODS[3]),
```
Comment (dgaubert) [inline lib/node/nodes/contour.js]: It's a pretty clever trick ;) and any javascripter should understand it but it's a bit hard to read. The idea was:

```
var  NEAREST_NEIGHBOR_METHOD = 'nearest_neighbor_method';
...
    method: Node.PARAM.NULLABLE(Node.PARAM.ENUM(
        NEAREST_NEIGHBOR_METHOD,
        .....
    ), QUANTILES_METHOD),
...
```
but If you are OK using `.apply()`, I'm OK too.

Comment (AbelVM) [inline lib/node/nodes/contour.js]: Hmmm... but that way I'd need to declare several variables. One for each string... This way I only need to maintain two arrays that are clearly defined.

But, if you prefer one var per string, I'll change it.

Comment (AbelVM) [inline lib/node/nodes/contour.js]: Plus: the `.indexOf()` trick of the template, stays clean and shinny

Comment (dgaubert) [inline lib/node/nodes/contour.js]: I have mixed feelings, :). If you are OK with that go ahead!

Comment (AbelVM) [inline lib/node/nodes/contour.js]: I'm soooo OK :stuck_out_tongue_winking_eye:

**Comment by AbelVM** [comment]
Boom!

**Pull request merged**